RavenDB-16913 OpenTelemetry - Metrics

maciejaszyk opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-16913

### Additional description

This PR introduces support for metrics provided via OpenTelemetry. See more details in the ticket and at https://github.com/ravendb/docs/pull/1862.
### Type of change

- [ ] Bug fix
- [ ] Regression bug fix
- [ ] Optimization
- [x] New feature

### How risky is the change?

- [x] Low 
- [ ] Moderate 
- [ ] High
- [ ] Not relevant

### Backward compatibility

- [x] Non breaking change
- [ ] Ensured. Please explain how has it been implemented?
- [ ] Breaking change
- [ ] Not relevant

### Is it platform specific issue?

- [ ] Yes. Please list the affected platforms.
- [x] No

### Documentation update

- [x] This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.
- [ ] No documentation update is needed 

### Testing by Contributor

- [ ] Tests have been added that prove the fix is effective or that the feature works
- [ ] Internal classes added to the test class (e.g. entity or index definition classes) have the lowest possible access modifier (preferable `private`) 
- [x] It has been verified by manual testing

### Testing by RavenDB QA team

- [ ] This change requires a special QA testing due to possible performance or resources usage implications (CPU, memory, IO). Please mark the issue on YouTrack using `QA Required` tag.
- [x] No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- [ ] Yes. Please list the affected features/subsystems and provide appropriate explanation
- [x
### UI work

- [ ] It requires further work in the Studio. Please mark the issue on YouTrack using `Studio Required` tag.
- [x] No UI work is needed

maciejaszyk committed:
RavenDB-16913 OpenTelemetry - initial work

maciejaszyk committed:
RavenDB-16913

- exposed more metrics
- exposed configuration to prefilter instrumentation
- exposing descriptions based on SNMP
- adding instrumentation table generator

ppekrol commented on src/Raven.Server/Config/Categories/MonitoringConfiguration.cs:
```diff
@@ -120,5 +122,129 @@ public sealed class SnmpConfiguration : ConfigurationCategory
+        [ConfigurationCategory(ConfigurationCategoryType.Monitoring)]
+        public sealed class OpenTelemetryConfiguration : ConfigurationCategory
+        {
+            [Description("Indicates if OpenTelemetry is enabled or not. Default: false")]
```
Default false, but 1 line down it is true :)

ppekrol commented on src/Raven.Server/Config/Categories/MonitoringConfiguration.cs:
```diff
@@ -120,5 +122,129 @@ public sealed class SnmpConfiguration : ConfigurationCategory
+            
+            [Description("Indicates if server-wide OpenTelemetry is enabled or not. Default: true")]
+            [DefaultValue(true)]
+            [ConfigurationEntry("Monitoring.OpenTelemetry.ServerWideEnabled", ConfigurationEntryScope.ServerWideOnly)]
```
Consider adding a dot between ServerWide and Enabled

maciejaszyk replied:
Done

ppekrol commented on src/Raven.Server/Config/Categories/MonitoringConfiguration.cs:
```diff
@@ -120,5 +122,129 @@ public sealed class SnmpConfiguration : ConfigurationCategory
+            
+            [Description("Indicates if database-wide OpenTelemetry is enabled or not. Default: true")]
+            [DefaultValue(true)]
+            [ConfigurationEntry("Monitoring.OpenTelemetry.DatabaseWideEnabled", ConfigurationEntryScope.ServerWideOnly)]
```
consider adding a dot

ppekrol commented on src/Raven.Server/Config/Categories/MonitoringConfiguration.cs:
```diff
@@ -120,5 +122,129 @@ public sealed class SnmpConfiguration : ConfigurationCategory
+            
+            [Description("Indicates if should use OpenTelementry protocol.")]
+            [DefaultValue(true)]
+            [ConfigurationEntry("Monitoring.OpenTelemetry.OpenTelemetryProtocolEnabled", ConfigurationEntryScope.ServerWideOnly)]
```
isn't the 2nd OpenTelemetry unnecessary?

maciejaszyk replied:
That's full name of protocol that OpenTelemetry introduced.

ppekrol commented on src/Raven.Server/Config/Categories/MonitoringConfiguration.cs:
```diff
@@ -120,5 +122,129 @@ public sealed class SnmpConfiguration : ConfigurationCategory
+            [Description("Indicates if should use OpenTelementry protocol.")]
+            [DefaultValue(true)]
+            [ConfigurationEntry("Monitoring.OpenTelemetry.OpenTelemetryProtocolEnabled", ConfigurationEntryScope.ServerWideOnly)]
+            public bool OltpExporter { get; set; }
```
I see a desync between property name and configuration entry key, is this by intent?

maciejaszyk replied:
no, solved

ppekrol commented on src/Raven.Server/Config/Categories/MonitoringConfiguration.cs:
```diff
@@ -120,5 +122,129 @@ public sealed class SnmpConfiguration : ConfigurationCategory
+            
+            [Description("Expose server storage instruments.")]
+            [DefaultValue(true)]
+            [ConfigurationEntry("Monitoring.OpenTelemetry.ServerStorage", ConfigurationEntryScope.ServerWideOnly)]
```
We need to work on configuration key conventions here, this should be ServerStorage.Enabled?

ppekrol commented on src/Raven.Server/Config/Categories/MonitoringConfiguration.cs:
```diff
@@ -120,5 +122,129 @@ public sealed class SnmpConfiguration : ConfigurationCategory
+            [Description("List of GC kinds to expose. Default: null (all).")]
+            [DefaultValue(null)]
+            [ConfigurationEntry("Monitoring.OpenTelemetry.GCKinds", ConfigurationEntryScope.ServerWideOnly)]
+            public string[] GCKinds { get; set; }
```
can use GCKind enum here maybe?

ppekrol commented on src/Raven.Server/Monitoring/OpenTelemetry/Constants.cs:
```diff
@@ -0,0 +1,200 @@
+
+    public static void Scanner()
+    {
+        return;
```
?

maciejaszyk replied:
Removed.

ppekrol commented on src/Raven.Server/Monitoring/OpenTelemetry/Constants.cs:
```diff
@@ -0,0 +1,200 @@
+            public const string ServerGcPromoted = Prefix + "promoted";
+            public const string ServerGcTotalAvailableMemory = Prefix + "totalavailablememory";
+            public const string ServerGcTotalCommitted = Prefix + "totalcommited";
+            public const string ServerGcLohSize = Prefix + "gclohsize";
```
is gc needed here in the name?

maciejaszyk replied:
Well, I've moved whole GC to meter `ravendb.serverwide.gc`

ppekrol commented on src/Raven.Server/Monitoring/OpenTelemetry/DatabaseWideMetrics.cs:
```diff
@@ -0,0 +1,289 @@
+    private readonly HashSet<string> _loadedIndexes = new(StringComparer.OrdinalIgnoreCase);
+
+    //Meter instances are universally shared across all indexes, as identification of indexes or databases occurs through measurement tags.
+    private static readonly Lazy<Meter> IndexesMeter = new(() => new Meter(Constants.IndexMeter, "1.0.0"));
```
what is this 1.0.0 here?

maciejaszyk replied:
OpenTelemetry gives you possibility to version meter implementation.
https://opentelemetry.io/docs/specs/otel/versioning-and-stability/

ppekrol commented on src/Raven.Server/Monitoring/OpenTelemetry/DatabaseWideMetrics.cs:
```diff
@@ -0,0 +1,289 @@
+
+    private void Attach(bool force)
+    {
+        if (Configuration.ExposedDatabases != null && Configuration.ExposedDatabases.Contains(_databaseName) == false)
```
this check needs to be case insensitive IMO?

ppekrol commented on src/Raven.Server/Monitoring/OpenTelemetry/DatabaseWideMetrics.cs:
```diff
@@ -0,0 +1,289 @@
+        if (force == false && _attached)
+            return;
+
+        Task.Factory.StartNew(async () =>
```
discard the task: `_ = Task.Factory...`

ppekrol commented on src/Raven.Server/Monitoring/OpenTelemetry/DatabaseWideMetrics.cs:
```diff
@@ -0,0 +1,289 @@
+                _attached = true;
+            }
+            catch (DatabaseDisabledException)
+            {
```
any other exception here will cause UnobservedTaskException?

ppekrol commented on src/Raven.Server/Monitoring/OpenTelemetry/DatabaseWideMetrics.cs:
```diff
@@ -0,0 +1,289 @@
+        if (change.Type != IndexChangeTypes.IndexAdded)
+            return;
+
+        Task.Factory.StartNew(async () =>
```
`_ = Task.Factory...`

ppekrol commented on src/Raven.Server/Monitoring/OpenTelemetry/DatabaseWideMetrics.cs:
```diff
@@ -0,0 +1,289 @@
+
+                RegisterIndexInstruments(change.Name);
+            }
+            finally
```
do we need ex handling here as well?

ppekrol commented on src/Raven.Server/Monitoring/OpenTelemetry/DatabaseWideMetrics.cs:
```diff
@@ -0,0 +1,289 @@
+        });
+    }
+
+    private void RegisterDatabaseStorageMetricMetric()
```
Metric twice?

maciejaszyk replied:
Unintended, fixed

ppekrol commented on src/Raven.Server/Monitoring/OpenTelemetry/IMetricInstrument.cs:
```diff
@@ -0,0 +1,15 @@
+
+public interface PlaceHolderReflection{}
+
+public interface IMetricInstrument<out TInstrumentValue>: PlaceHolderReflection
```
formatting

maciejaszyk replied:
Done

ppekrol commented on src/Raven.Server/Monitoring/OpenTelemetry/MetricsManager.cs:
```diff
@@ -0,0 +1,153 @@
+
+            _server.ServerStore.DatabasesLandlord.OnDatabaseLoaded += AddDatabasesIfNecessary;
+        }
+        catch (Exception e)
```
compiler warning

ppekrol commented on src/Raven.Server/Monitoring/OpenTelemetry/MetricsManager.cs:
```diff
@@ -0,0 +1,153 @@
+        }
+        catch (Exception e)
+        {
+            Debugger.Launch();
```
remember about this

maciejaszyk replied:
removed

ppekrol commented on src/Raven.Server/Monitoring/OpenTelemetry/MetricsManager.cs:
```diff
@@ -0,0 +1,153 @@
+        AsyncHelpers.RunSync(AddDatabases);
+    }
+
+    private async Task AddDatabases()
```
AddDatabasesAsync

ppekrol commented on src/Raven.Server/Monitoring/OpenTelemetry/MetricsManager.cs:
```diff
@@ -0,0 +1,153 @@
+        if (_server.Configuration.Monitoring.OpenTelemetry.Databases == false)
+            return;
+
+        await _locker.WaitAsync();
```
beside this, I do not see any async code here and IIRC SemaphoreSlim has sync version of that. Do we want to take leverage of it?

ppekrol commented on src/Raven.Server/ServerWide/ServerStore.cs:
```diff
@@ -612,7 +612,7 @@ public void RequestSnapshot()
         }
 
-        public void Initialize()
+        public void Initialize_Phase_1()
```
PreInitialize

maciejaszyk replied:
fixed

ppekrol commented on src/Raven.Server/ServerWide/ServerStore.cs:
```diff
@@ -854,8 +853,11 @@ public void Initialize()
+        }
 
-
+        public void Initialize_Phase_2()
```
Initialize

maciejaszyk replied:
fixed

maciejaszyk committed:
RavenDB-16913: Unifying configuration options, creating category meters, and removing filtration at the instrument level.

maciejaszyk committed:
RavenDB-16913 Removing metrices per database.

maciejaszyk committed:
RavenDB-16913
- Introducing CPU credit meter
- Unifying configuration options

maciejaszyk committed:
RavenDB-16913 OpenTelemetry - SnmpOids description scanner fixes after rabase.

maciejaszyk committed:
RavenDB-16913 New description generator, removing leftovers, typos fixes

maciejaszyk commented on src/Raven.Server/Monitoring/OpenTelemetry/MetricsManager.cs:
```diff
@@ -0,0 +1,84 @@
+            if (activate)
+            {
+                if (_serverMetrics != null)
+                    throw new InvalidOperationException("Cannot start SNMP Engine because it is already activated. Should not happen!");
```
...snmp

maciejaszyk committed:
RavenDB-16913 Updating OpenTelemetry packages &  adding configuration for OtlpExporter & cleanup

arekpalinski commented on src/Raven.Server/Config/Categories/MonitoringConfiguration.cs:
```diff
@@ -30,18 +32,114 @@ public sealed class MonitoringConfiguration : ConfigurationCategory
+            [ConfigurationEntry("Monitoring.OpenTelemetry.Enabled", ConfigurationEntryScope.ServerWideOnly)]
+            public bool Enabled { get; set; }
+            
+            [Description("Indicates if server-wide OpenTelemetry is enabled or not. Default: true")]
```
How does it differ from the above option? I think we already talked about that and this option is supposed to enable / disable the server wide metrics / instruments, right? If yes then the description needs to be adjusted (config option name as well?)

maciejaszyk replied:
Well, when you want to expose dotnet instruments but don't want to expose server meters this could be useful.

arekpalinski replied:
Actually I didn't ask about the option itself but rather about the description

arekpalinski replied:
Below options has "... metrics are enabled or not". Note "metrics" while here you say "Open Telemetry is enabled"

maciejaszyk replied:
`Indicates if RavenDB's OpenTelemetry metrics are enabled or not. Default: true`?

arekpalinski commented on src/Raven.Server/Config/Categories/MonitoringConfiguration.cs:
```diff
@@ -30,18 +32,114 @@ public sealed class MonitoringConfiguration : ConfigurationCategory
+            
+            [Description("Expose instruments related to server storage.")]
+            [DefaultValue(true)]
+            [ConfigurationEntry("Monitoring.OpenTelemetry.ServerWide.Storage.Enabled", ConfigurationEntryScope.ServerWideOnly)]
```
In above config options names you use  ".Metrics.XXXXInstrumentation." e.g. Monitoring.OpenTelemetry.Metrics.RuntimeInstrumentation.Enabled. 

Wouldn't it make sense to apply it here (and below as well)? Can we make the naming consistent?

arekpalinski replied:
Also instead of `ServerWide` can we use just `Server`?

maciejaszyk replied:
I like the idea. 
I would go with:
e.g.
`Monitoring.OpenTelemetry.Meters.Server.Storage.Enabled`

maciejaszyk replied:
Also regarding this comment: https://github.com/ravendb/ravendb/pull/18589#discussion_r1658283298
For built-in:
`Indicates if AspNetCore metric is enabled or not. Default: true`
`Indicates if Runtime metric is enabled or not. Default: true`
And for ours:
I would suggest something like that:
`"Expose metric related to [NAME]."`

arekpalinski commented on src/Raven.Client/DocumentationUrls.cs:
```diff
@@ -15,6 +15,18 @@ internal static class Querying
             public const string HighlightQueryResults = nameof(HighlightQueryResults);
 
+            /// <remarks><seealso href="https://ravendb.net/docs/article-page/6.0/Csharp/client-api/session/querying/debugging/query-timings"/></remarks>
+            public const string QueryTimings = nameof(QueryTimings);
```
Why do we have those changes? It was added in dedicated 6.0 PR

arekpalinski commented on src/Raven.Client/Exceptions/Documents/Indexes/IndexDoesNotExistException.cs:
```diff
@@ -41,5 +41,10 @@ public static IndexDoesNotExistException ThrowFor(string indexName)
             throw new IndexDoesNotExistException($"There is no index with '{indexName}' name.");
         }
+
+        public static IndexDoesNotExistException ThrowForAuto(string indexName)
```
Is this change relevant for this PR?

arekpalinski commented on src/Raven.Server/Monitoring/Snmp/Objects/Server/1.6/1.6.11/ServerGcBase.cs:
```diff
@@ -9,12 +10,14 @@ public abstract class ServerGcBase<TData> : ScalarObjectBase<TData>
     {
         private readonly MetricCacher _metricCacher;
         private readonly string _cacheKey;
+        protected KeyValuePair<string, object?> MeasurementTag;
```
Please check the warning

maciejaszyk replied:
Removed nullable.

arekpalinski commented on src/Raven.Server/ServerWide/ClusterStateMachine.cs:
```diff
@@ -4903,6 +4903,15 @@ public unsafe bool IsReplicationCertificateByPublicKeyPinningHash(ClusterOperati
+            using (context.OpenReadTransaction())
+                nodeTag = RachisConsensus.ReadNodeTag(context);
+
+            return nodeTag != RachisConsensus.InitialTag;
```
Why do we compare the value we read with `RachisConsensus.InitialTag;`?

maciejaszyk replied:
To track the instance from which instruments come from, we pass a node tag as `instanceId`. However, if you have a fresh RavenDB instance and the cluster tag is not yet available, we skip the OpenTelemetry initialization.

arekpalinski replied:
Maybe we should log that information? That we skipped that

maciejaszyk replied:
Added

ppekrol commented on src/Corax/Constants.cs:
```diff
@@ -1,5 +1,6 @@
 using System;
 using System.Runtime.CompilerServices;
+using System.Text;
```
remove?

ppekrol commented on src/Corax/Constants.cs:
```diff
@@ -34,6 +35,7 @@ static Constants()
         {
             using (StorageEnvironment.GetStaticContext(out var ctx))
             {
+                
```
remove?

ppekrol commented on src/Raven.Server/Commercial/LicenseManager.cs:
```diff
@@ -1449,7 +1449,7 @@ public void AssertCanDelayReplication(TimeSpan delayReplicationFor)
         }
 
-        public void AssertCanUseDocumentsCompression(DocumentsCompressionConfiguration documentsCompression)
+        public void  AssertCanUseDocumentsCompression(DocumentsCompressionConfiguration documentsCompression)
```
formatting?

ppekrol commented on src/Raven.Server/Commercial/LicenseManager.cs:
```diff
@@ -1605,6 +1605,33 @@ public void AssertCanAddReadOnlyCertificates(CertificateDefinition certificate)
+                    return true;
+                else
+                {
+                    const string details = "Your license allows you to run OpenTelemetry meters but in order to run on meters you've to manually restart the RavenDB process.";
```
Not sure I follow the message? the part 'run on meters' is confusing for me

ppekrol commented on src/Raven.Server/Monitoring/OpenTelemetry/MetricsManager.cs:
```diff
@@ -0,0 +1,84 @@
+        try
+        {
+            var activate = _server.ServerStore.LicenseManager.CanUseOpenTelemetryMonitoring(withNotification: true, startUp: false);
+            if (activate)
```
you are not taking into account here that you might already ran it, this even can fire few times during single run, in SNMP we are checking the state and if it was already active then we are not doing anything

maciejaszyk replied:
Added a boolean value as an indicator to show whether we registered metrics (and skip if we already did it).

ppekrol commented on src/Raven.Server/Monitoring/OpenTelemetry/MetricsManager.cs:
```diff
@@ -0,0 +1,84 @@
+
+        try
+        {
+            var activate = _server.ServerStore.LicenseManager.CanUseOpenTelemetryMonitoring(withNotification: true, startUp: true);
```
check my prev comment regarding activation

ppekrol commented on src/Raven.Server/Monitoring/Snmp/SnmpWatcher.cs:
```diff
@@ -89,9 +89,9 @@ private void OnLicenseChanged()
 
         public void Execute()
-        {
+        { 
```
formatting?

ppekrol commented on src/Raven.Server/Monitoring/Snmp/SnmpWatcher.cs:
```diff
@@ -89,9 +89,9 @@ private void OnLicenseChanged()
+        { 
             if (_server.Configuration.Monitoring.Snmp.Enabled == false)
-                return;
+                 return;
```
formatting?

arekpalinski commented on src/Raven.Server/Config/Categories/MonitoringConfiguration.cs:
```diff
@@ -30,18 +32,114 @@ public sealed class MonitoringConfiguration : ConfigurationCategory
+            
+            [Description("Indicates if metrics should be exported with the OpenTelemetry protocol.")]
+            [DefaultValue(true)]
+            [ConfigurationEntry("Monitoring.OpenTelemetry.OpenTelemetryProtocol.Enabled", ConfigurationEntryScope.ServerWideOnly)]
```
Why having "OpenTelemetry" repeated. 
```suggestion
            [ConfigurationEntry("Monitoring.OpenTelemetry.Protocol.Enabled", ConfigurationEntryScope.ServerWideOnly)]
```

maciejaszyk replied:
because `OpenTelemetryProtocol` is official name of their protocol: https://opentelemetry.io/docs/specs/otlp/

maciejaszyk committed:
RavenDB-16913
- PR comments
- overload for ReadNodeTag will use ReadTree instead of CreateTree, since it can be called from a read transaction
- naming conventions
- activation flag inside the OpenTelemetry manager will prevent over-creating meters.

arekpalinski commented:
@maciejaszyk please fill `Additional description` section

arekpalinski commented on src/Raven.Server/Config/Categories/MonitoringConfiguration.cs:
```diff
@@ -30,18 +32,114 @@ public sealed class MonitoringConfiguration : ConfigurationCategory
+            [Description("Indicates if server-wide OpenTelemetry is enabled or not. Default: true")]
+            [DefaultValue(true)]
+            [ConfigurationEntry("Monitoring.OpenTelemetry.Meters.Server.Enabled", ConfigurationEntryScope.ServerWideOnly)]
+            public bool ServerWideEnabled { get; set; }
```
You changes the config option ServerWide -> Server but the property is still ServerWide

arekpalinski replied:
Below options have `".Instrumentation."` should we add it here as well for consistency or maybe drop it in below options?

maciejaszyk replied:
Changed

maciejaszyk committed:
RavenDB-16913
- log information why opentelemetry is uninitialized when NodeTag is InitialTag
- naming conventions

maciejaszyk committed:
RavenDB-16913 Remove unused package.

ppekrol commented on src/Raven.Server/Config/Categories/MonitoringConfiguration.cs:
```diff
@@ -30,18 +32,114 @@ public sealed class MonitoringConfiguration : ConfigurationCategory
+            public bool ServerMetersEnabled { get; set; }
+            
+            [Description("Indicates if AspNetCore metric is enabled or not. Default: true")]
+            [DefaultValue(true)]
```
false?

ppekrol commented on src/Raven.Server/Config/Categories/MonitoringConfiguration.cs:
```diff
@@ -30,18 +32,114 @@ public sealed class MonitoringConfiguration : ConfigurationCategory
+            public bool AspNetCoreInstrumentationMetersEnabled { get; set; }
+            
+            [Description("Indicates if Runtime metric is enabled or not. Default: true")]
+            [DefaultValue(true)]
```
false?

ppekrol commented on src/Raven.Server/Config/Categories/MonitoringConfiguration.cs:
```diff
@@ -30,18 +32,114 @@ public sealed class MonitoringConfiguration : ConfigurationCategory
+            [ConfigurationEntry("Monitoring.OpenTelemetry.Meters.Server.GC.Enabled", ConfigurationEntryScope.ServerWideOnly)]
+            public bool GcEnabled { get; set; }
+            
+            [Description("Expose metric related to general.")]
```
not a good description I think?

ppekrol commented on src/Raven.Server/RavenServer.cs:
```diff
@@ -411,6 +427,93 @@ void ConfigureKestrel(KestrelServerOptions options)
+
+            if (ClusterStateMachine.TryReadNodeTag(ServerStore, out var nodeTag) == false)
+            {
+                Logger.Info("OpenTelemetry monitoring requires the node tag of the server for initialization; however, it is still unavailable. Therefore, OpenTelemetry initialization is skipped.");
```
Logger.IsInfoEnabled?

maciejaszyk committed:
RavenDB-16913 Default settings / log only when logging is enabled.
